### PR TITLE
fix: Refactor handling of empty chart data in ChartWidget

### LINKED ISSUE #37008: Bug: ChartWidget sentry issue. TypeError: Cannot convert undefined or null to object
Sentry Issue: [APPSMITH-A97](https://appsmith.sentry.io/issues/6011588612/?referrer=github_integration)

```
TypeError: Cannot convert undefined or null to object
  at Function.keys (<anonymous>)
  at emptyChartData (widgets/ChartWidget/widget/index.tsx:54:18)
  at <object>.value (widgets/ChartWidget/widget/index.tsx:237:9)
  at <object>.value (widgets/BaseWidget.tsx:383:17)
  at idleCallbackId (widgets/withLazyRender.tsx:36:19)
...
(14 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description
* This pull request handles empty chart dAata more efficiently. 
* lso updates tests to ensure this is not overlooked again.

These changes ensure that the ChartWidget can handle scenarios where the chart data is null or undefined, and that the tests accurately reflect this behavior.

Fixes #37008
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Chart"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11473049534>
> Commit: 08602d3a0658b1753d4d377ce9673379ab234d3f
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11473049534&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Chart`
> Spec:
> <hr>Wed, 23 Oct 2024 04:52:48 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved data handling in the ChartWidget to prevent runtime errors when chart data is undefined.
  
- **Tests**
	- Added a new test case to verify the behavior of the emptyChartData function when series data is null or undefined, enhancing test coverage. 

- **Documentation**
	- Updated interface to include an optional property for handling data point clicks in the ChartWidget.

### CODE REVIEW

File: app/client/src/widgets/ChartWidget/widget/index.test.ts [hunk 1]
```diff
@@ -72,6 +72,15 @@ describe("emptyChartData", () => {
       expect(emptyChartData(props)).toEqual(true);
     });
 
+    it("returns true if each series is null or undefined", () => {
+      const props = JSON.parse(JSON.stringify(basicEChartProps));
```
Comment (jsartisan) [inline app/client/src/widgets/ChartWidget/widget/index.test.ts:76]: I know you are just following what the previous developer did in this test, why are we doing stringify and parse the JS object? 🤔 Just curious about it...What's the point?

File: app/client/src/widgets/ChartWidget/widget/index.tsx [hunk 2]
```diff
@@ -51,7 +51,10 @@ export const emptyChartData = (props: ChartWidgetProps) => {
     const builder = new EChartsDatasetBuilder(props.chartType, props.chartData);
 
     for (const seriesID in builder.filteredChartData) {
-      if (Object.keys(props.chartData[seriesID].data).length > 0) {
+      if (
+        props.chartData[seriesID].data &&
```
Comment (jsartisan) [inline app/client/src/widgets/ChartWidget/widget/index.tsx:55]: Maybe simply do `Array.isArray` here instead? it will also make sure this is an array.

**Review by jsartisan — DISMISSED** [review]
LGMT, left 2 tiny comments...

File: app/client/src/widgets/ChartWidget/widget/index.test.ts (hunk 1 above)
Comment (rahulbarwal) [reply app/client/src/widgets/ChartWidget/widget/index.test.ts:76]: Basically creating a (deep)clone of the object. We change some nested properties of this object in the spec.

File: app/client/src/widgets/ChartWidget/widget/index.tsx (hunk 2 above)
Comment (rahulbarwal) [reply app/client/src/widgets/ChartWidget/widget/index.tsx:55]: Actually this is supposed to be an object. but I see your point, let me accommodate this.

Comment (rahulbarwal) [reply app/client/src/widgets/ChartWidget/widget/index.tsx:55]: Ignore my last comment, this makes sense.

Comment (rahulbarwal) [reply app/client/src/widgets/ChartWidget/widget/index.tsx]: Done.

**Review by jsartisan — APPROVED** [review]

**Pull request merged**